### PR TITLE
feat: initial support for Arris devices (Vodafone Station TG3442DE)

### PR DESCRIPTION
**Pull request by phrxmd**
This adds basic support for the German TG3442DE Vodafone stations. Internally this is a repackaged Arris TG3442 cable router.

The Arris firmware is an odd beast. The station runs `lighthttpd` and PHP.  It does not provide a "proper" JSON API, but it provides its own API that serves structured JavaScript data and JSON objects through a PHP endpoint in the backend of the web interface. 

The web frontend consists of a single HTML page `/index.php` with an empty body and a lot of JavaScript.  When accessing this in the browser,  an Arris formatter and a Vodafone theming script build a DOM for a pretty user interface.  Different pages are generated through URLs of the type `/index.php?status_status`.  The actual data comes from a backend API that provides a number of topic-specific endpoints `/php/*_data.php`.  These are PHP pages that serve up a `<script>` tag with a list of JavaScript variables and JSON objects, like this (from `/php/status_status_data.php`):

```
<script type="text/javascript">

var js_SerialNumber = '1234567890';
var js_FWVersion = 'AR01.04.046.25_091322_7244.PC20.10.X1';
var js_HWTypeVersion = '8'; 
[...]
var js_ethernet_port4_status  = 'Active';
var js_ethernet_port4_bitrate = '1 Gbps';
[...]

</script>
```

We can use the PHP endpoints to access the API, as long as we can obtain and maintain a login session. This is a bit tricky because the Arris firmware is picky with things like headers, and because it's an odd system, with lots of semi-secure encryption going on.  But it gives us structured information with well-defined keys and is perfectly workable.  It's certainly much preferable to scraping the web UI's HTML, which is only generated in the browser anyway and hard to access, short of a client-side JS implementation.

There are a lot of values exposed through JS variables; I intend to eventually add support for most of them. Next up is DOCSIS 3.1 link quality monitoring.

The code is based on : 
- https://github.com/nox-x/TG3442DE-Teardown/
- https://github.com/heeplr/munin-arris/ (Public domain)
- https://github.com/totev/vodafone-station-cli (MIT license)
- https://github.com/fsck-block/arris-tg3442de-exporter/ (Apache 2.0 license)

### CODE REVIEW

File: src/aiovodafone/api.py [hunk 1]
```diff
@@ -65,14 +73,21 @@ async def get_device_type(
                 response_json = await response.json()
                 if "data" in response_json and "ModelName" in response_json["data"]:
                     return DeviceType.TECHNICOLOR
-        async with session.get(
-            f"https://{host}/login.html", headers=HEADERS, ssl=False
-        ) as response:
+        async with session.get(f"http://{host}/login.html") as response:
```
Comment (chemelli74) [inline src/aiovodafone/api.py:76]:
```suggestion
        async with session.get(
            f"https://{host}/login.html", headers=HEADERS, ssl=False
        ) as response:
```
```

File: src/aiovodafone/api.py [hunk 2]
```diff
@@ -65,14 +73,21 @@ async def get_device_type(
...
             if response.status == 200:
                 # To identify the Sercomm devices before the login
                 # There's no other sure way to identify a Sercomm device without login
                 if "var csrf_token = " in await response.text():
                     return DeviceType.SERCOMM
+        async with session.get(
+            f"http://{host}/index.php",
+            headers=HEADERS.update(
+                {"Accept:": "text/html,application/xhtml+xml,application/xml"}
+            ),
```
Comment (chemelli74) [inline src/aiovodafone/api.py:86]:
```suggestion
            headers=HEADERS
```

Comment (chemelli74) [reply src/aiovodafone/api.py:86]: Please add the additional header to const.py -> HEADERS

**Review by chemelli74 — CHANGES_REQUESTED** [review]
Left a couple of comments

File: src/aiovodafone/api.py (hunk 1 above)
Comment (phrxmd) [reply src/aiovodafone/api.py]: Done

File: src/aiovodafone/api.py (hunk 2 above)
Comment (phrxmd) [reply src/aiovodafone/api.py]: Done

File: src/aiovodafone/api.py [hunk 3]
```diff
@@ -164,6 +176,449 @@ async def logout(self) -> None:
...
+            """Initialize Arris endpoint.
+
+            Args:
+                api (VodafoneStationCommonApi): the outer class instance used for session management
+                name (str): descriptive name of the endpoint for logging
+                page (str): the endpoint address in the Vodafone Station web UI
+                vars ({str: str}): mapping of sensor names to Vodafone Station JavaScript variable names
+                jsons ({str: str}): mapping of sensor names to Vodafone Station JSON object names
+            """
+            _LOGGER.debug(f"Initializing {name!r} endpoint for retrieval from {page!s}")
```
Comment (chemelli74) [inline src/aiovodafone/api.py:255]:
```suggestion
            _LOGGER.debug("Initializing%s endpoint for retrieval from %s", name, page)
```

File: src/aiovodafone/api.py [hunk 4]
```diff
@@ -164,6 +176,449 @@ async def logout(self) -> None:
...
+            self.vars = vars
+            self.jsons = jsons
+            # Build empty dictionary for all values we are expected to generate
+            self.data = dict.fromkeys([*self.vars] + [*self.jsons])
+
+        async def retrieve(self) -> None:
+            """Retrieve JavaScript variables and JSON data."""
+            response = await self.api._get_page_result(self.page)
+            raw_data = await response.text()
+            # _LOGGER.debug(f"Here's our raw page <{self.page}>:\n{raw_data}")
```
Comment (chemelli74) [inline src/aiovodafone/api.py:268]:
```suggestion
```

File: src/aiovodafone/api.py [hunk 5]
```diff
@@ -164,6 +176,449 @@ async def logout(self) -> None:
...
+            for _var in self.vars.keys():
+                # Single values (`var js_SomeVar = 'value'`).
+                raw = await self._search(
```
Comment (chemelli74) [inline src/aiovodafone/api.py:271]: Not sure is allowed. I fear it can be seen as web scraping

File: src/aiovodafone/api.py [hunk 6]
```diff
@@ -164,6 +176,449 @@ async def logout(self) -> None:
...
+                    r".*var " + self.vars[_var] + r" = '(.*)';.*", raw_data
+                )
+                self.data[_var] = raw[0]
+            for _json in self.jsons.keys():
+                # JSON objects (`json_SomeData = {...}`).
+                raw = await self._search(
```
Comment (chemelli74) [inline src/aiovodafone/api.py:277]: Same as above

File: src/aiovodafone/api.py [hunk 7]
```diff
@@ -164,6 +176,449 @@ async def logout(self) -> None:
...
+
+        Args:
+            ciphertext (bytes): text to be decrypted
+            key (bytes): encryption key generated from the credentials and salt
+            iv (bytes): iv given by the web UI
+
+        Returns:
+            bytes: the decrypted plaintext
+        """
+        _LOGGER.debug(f"Decrypt ciphertext {ciphertext!r}")
```
Comment (chemelli74) [inline src/aiovodafone/api.py:361]:
```suggestion
        _LOGGER.debug("Decrypt ciphertext %s", ciphertext)
```

File: src/aiovodafone/api.py [hunk 8]
```diff
@@ -164,6 +176,449 @@ async def logout(self) -> None:
...
+            raise SessionLost("Lost existing Arris firmware login session")
+
+    async def login(self) -> bool:
+        """
+        Login to Vodafone Station with Arris firmware.
+
+        Emulate the station's convoluted login process with
+        encrypted negotiation between browser and firmware.
+        """
+        _LOGGER.debug(f"Attempting to log into Arris Vodafone Station at {self.host}")
```
Comment (chemelli74) [inline src/aiovodafone/api.py:382]:
```suggestion
        _LOGGER.debug("Attempting to log into Arris Vodafone Station at %s", self.host)
```

**Review by chemelli74 — CHANGES_REQUESTED** [review]
Please review all log messages, I suggested a few changes but all need to be rechecked.

Also not sure if the way you search keys is fine or not.

File: src/aiovodafone/api.py (hunk 3 above)
Comment (phrxmd) [reply src/aiovodafone/api.py:255]: ok

File: src/aiovodafone/api.py (hunk 4 above)
Comment (phrxmd) [reply src/aiovodafone/api.py:268]: ok

File: src/aiovodafone/api.py (hunk 7 above)
Comment (phrxmd) [reply src/aiovodafone/api.py]: ok

**Comment by phrxmd** [comment]
Superseded by #89.

**Pull request closed without merging**